File: microservices/microservice.js
const { faker } = require("@faker-js/faker");
const winston = require("winston");
require("winston-daily-rotate-file");
const fs = require('fs');
const path = require('path');

// Ensure a writable logs directory exists (relative to project root)
const logsDir = path.join(process.cwd(), 'logs');
try {
  fs.mkdirSync(logsDir, { recursive: true });
} catch (err) {
  // If creation fails, leave it to the transport to report an error later
  // (e.g., read-only file system). We avoid throwing here so the app can
  // surface the original error with context.
}

// Configure Daily Rotate File transport (writes into the project's ./logs)
const fileRotateTransport = new winston.transports.DailyRotateFile({
  filename: path.join(
    logsDir,
    `${process.env.SERVICE_NAME || "service"}-%DATE%.log`
  ),
  datePattern: "YYYY-MM-DD-HH",
  maxsize: "20m",
  maxFiles: "5",
  format: winston.format.json(),
});

// Create Winston logger
const logger = winston.createLogger({
  level: "debug",
  /*   level: "info",
    format: winston.format.combine(
      winston.format.timestamp(),
      winston.format.json()
    ), */
  transports: [
    fileRotateTransport,
    new winston.transports.Console({
      format: winston.format.simple()
    })
  ]
});

class LogGenerator {
  constructor() {
    if (process.env.SERVICE_NAME) {
      this.serviceNames = [process.env.SERVICE_NAME];
    } else {
      this.serviceNames = [
        "web-service",
        "auth-service",
        "user-service",
        "order-service",
        "inventory-service",
        "payment-service",
      ];
    }
    this._timer = null;
    this._stopped = false;
  }

  // Start generating logs periodically. This returns a Promise that only
  // resolves when the generator is stopped or the process exits.
  async start(intervalMs = 1000) {
    if (this._timer) return; // already started

    this._stopped = false;
    this._timer = setInterval(() => {
      const service = this.serviceNames[
        Math.floor(Math.random() * this.serviceNames.length)
      ];
      const level = ["info", "warn", "error", "debug"][
        Math.floor(Math.random() * 4)
      ];
      const message = faker.hacker.phrase();

      // Use the Winston logger defined above
      logger.log({ level, message, service, timestamp: new Date().toISOString() });
    }, intervalMs);

    // Return a promise that never resolves until stopped so `await start()` keeps the process alive
    return new Promise((resolve) => {
      this._resolve = resolve;
    });
  }

  stop() {
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
    }
    this._stopped = true;
    if (this._resolve) this._resolve();
  }
}

async function main() {
  const logGenerator = new LogGenerator();
  try {
    await logGenerator.start();
  } catch (error) {
    console.error("Error running simulation:", error);
    process.exit(1);
  }
}

// Handle graceful shutdown
process.on("SIGTERM", () => {
  console.log("Received SIGTERM. Shutting down gracefully...");
  process.exit(0);
});

process.on("SIGINT", () => {
  console.log("Received SIGINT. Shutting down gracefully...");
  process.exit(0);
});

//Run the simulation
if (require.main === module) {
  main().catch(console.error);
}

File: microservices/microservice.py
import os
import time
from datetime import datetime
import random
from pathlib import Path
import logging
from logging.handlers import TimedRotatingFileHandler
from faker import Faker

# Initialize Faker
faker = Faker()

# Ensure logs directory exists
LOGS_DIR = Path(__file__).parent / 'logs'
LOGS_DIR.mkdir(exist_ok=True)

# Configure logger
def setup_logger():
    logger = logging.getLogger('microservice')
    logger.setLevel(logging.DEBUG)
    
    # Create formatter
    formatter = logging.Formatter(
        '{"timestamp": "%(asctime)s", "level": "%(levelname)s", "service": "%(service)s", "message": "%(message)s"}'
    )
    
    # Console handler
    console_handler = logging.StreamHandler()
    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)
    
    # Rotating file handler
    service_name = os.getenv('SERVICE_NAME', 'service')
    current_time = datetime.now().strftime('%Y-%m-%d-%H')
    base_filename = LOGS_DIR / f"{service_name}-{current_time}.log"
    file_handler = TimedRotatingFileHandler(
        filename=base_filename,
        when='H',  # Rotate every hour
        interval=1,
        backupCount=5,  # Keep 5 backup files
        encoding='utf-8'
    )
    # Custom namer function to maintain the date format in rotated files
    def namer(default_name):
        # Extract the base name without extension
        base = os.path.basename(default_name)
        if base.endswith(".log"):
            # Remove .log from the end
            base = base[:-4]
        # Get the current hour's timestamp
        timestamp = datetime.now().strftime('%Y-%m-%d-%H')
        # Construct the new name
        return str(LOGS_DIR / f"{service_name}-{timestamp}.log")
    
    file_handler.namer = namer
    file_handler.setFormatter(formatter)
    logger.addHandler(file_handler)
    
    return logger

class LogGenerator:
    def __init__(self):
        self.service_names = (
            [os.getenv('SERVICE_NAME')] if os.getenv('SERVICE_NAME') else [
                "web-service",
                "auth-service",
                "user-service",
                "order-service",
                "inventory-service",
                "payment-service",
            ]
        )
        self._stopped = False
        self.logger = setup_logger()

    def generate_log(self):
        """Generate a single log entry"""
        service = random.choice(self.service_names)
        level = random.choice(['info', 'warning', 'error', 'debug'])
        message = faker.catch_phrase()
        
        # Get the logging method corresponding to the level
        log_method = getattr(self.logger, level)
        
        # Add extra={'service': service} to include service in the log record
        log_method(message, extra={'service': service})

    def start(self, interval_ms=1000):
        """Start generating logs periodically"""
        interval_sec = interval_ms / 1000
        self._stopped = False
        
        try:
            while not self._stopped:
                self.generate_log()
                time.sleep(interval_sec)
        except KeyboardInterrupt:
            self.stop()

    def stop(self):
        """Stop generating logs"""
        self._stopped = True

def main():
    log_generator = LogGenerator()
    try:
        log_generator.start()
    except KeyboardInterrupt:
        log_generator.stop()

if __name__ == "__main__":
    main()